Unify search boxes

denyskon opened the pull request:
Unify all but a few search boxes to use uniform style, uniform translations and shared templates where possible.
Remove a few duplicated search templates, e. g. code search.

<details><summary>Example after screenshots:</summary>

![grafik](https://github.com/go-gitea/gitea/assets/47871822/e20e7d6b-c6be-4a47-b132-672766f41421)

![grafik](https://github.com/go-gitea/gitea/assets/47871822/d5b11b9c-c12f-4a29-8fb0-24e5aa511d18)

![grafik](https://github.com/go-gitea/gitea/assets/47871822/d86bb444-36c7-426d-9cf1-c634963dffb1)

![grafik](https://github.com/go-gitea/gitea/assets/47871822/a76c0319-0518-484a-a840-563d02b61198)

</details>


Also includes #29700 

Co-authored-by: 6543 <6543@obermui.de>

denyskon commented on options/locale/locale_en-US.ini:
```diff
@@ -1989,17 +1986,6 @@ contributors.contribution_type.commits = Commits
 contributors.contribution_type.additions = Additions
 contributors.contribution_type.deletions = Deletions
 
-search = Search
-search.search_repo = Search repository
-search.type.tooltip = Search type
-search.fuzzy = Fuzzy
-search.fuzzy.tooltip = Include results that also matches the search term closely
-search.match = Match
-search.match.tooltip = Include only results that matches the exact search term
-search.results = Search results for "%s" in <a href="%s">%s</a>
-search.code_no_results = No source code matching your search term found.
-search.code_search_unavailable = Currently code search is not available. Please contact your site administrator.
```
How did that even happen....? Exact duplicate of line 340+ :laughing:

delvh commented on options/locale/locale_en-US.ini:
```diff
@@ -142,6 +142,19 @@ confirm_delete_selected = Confirm to delete all selected items?
+fuzzy_tooltip = Include results that also match the search term closely
+match = Match
+match_tooltip = Include only results that match the exact search term
+no_results = No %s matching your search term found.
```
Not a good idea to deduplicate this term.
Or are you using it as `ctx.Locale.Tr("…no_results", ctx.Locale.Tr("…repo"))` for example?

delvh replied:
Yep, you did exactly that…

lafriks replied:
This is very tricky, while it works in English it will not work in other languages (word casing, word forms etc, for example in Latvian each noun has 7 forms for each singular and multiple cases with different endings depending on the sentence where it's used), so explore labels definitely cannot be used here

denyskon replied:
@lafriks Maybe simply do "Search..." everywhere where it's obvious?

lafriks replied:
I would not mind from my perspective but see @fnetX comment and I agree that it's better to have specific placeholder for accessibility

delvh commented on templates/repo/search.tmpl:
```diff
@@ -20,12 +20,9 @@
+				<p>{{ctx.Locale.Tr "search.unavailable" (ctx.Locale.Tr "search.kind.code")}}</p>
 			</div>
 		{{else if .Keyword}}
-			<h3>
-				{{ctx.Locale.Tr "repo.search.results" .Keyword .RepoLink .RepoName}}
-			</h3>
```
?

denyskon replied:
I don't see a reason for this to be there. In every other case we simply output the search results, but with code search we first print this header containing the search term (already visible in the search field above) and the repo (should be obvious)

lafriks reviewed (changes requested):
As per my comment

lafriks commented on options/locale/locale_en-US.ini:
```diff
@@ -142,6 +142,19 @@ confirm_delete_selected = Confirm to delete all selected items?
 
+[search]
+search = Search
+search_kind = Search %s...
```
Same here

denyskon commented:
I will refactor search templates as a whole.....

fnetX commented:
Hi there, I'm just checking by … I am currently also refactoring some search boxes for Forgejo for accessibility, and my work is probably exactly the opposite of this pull. I cannot block your decisions, but please consider the following.

> Maybe simply do "Search..." everywhere where it's obvious?

The meaning of search boxes is usually obvious to people with intact eye vision, depending on where a search box is placed. A search box at the top of the page conveys a global search, a search box just on top of a repo sidebar conveys it searches the repos. However, if the visual layout of the page is not available (and the semantic representation does not help either – mostly the case in this codebase), a screenreader reading "Search" to you leaves you in the dark about what you are about to search.

The simplest solution would be to add a "Search x" string for every search box including an aria-label or labelled-by element, because a placeholder on its own does not satisfy accessibility requirements. If you really want, you can keep the Search ... placeholder, but in any case I recommend to keep the possibility to specify the exact meaning of a search box.

And in localization, there is also often good reason to have duplicate terms, because grammar rules differ in many languages. Reusing strings is often done at the level of localization tools itself, and while I cannot say anything about the proprietary Crowdin, at least Weblate has a "translation memory" which offers translators access to similar strings – still keeping the ability to adapt if needed.

denyskon commented:
@fnetX @lafriks @delvh

I now refactored the whole bunch of search templates, including using shared templates where possible. I also added a corresponding & uniform "Search *... (repos/orgs/users/*" translation key for most search boxes to ensure a11y.

6543 commented on templates/shared/search/input.tmpl:
```diff
@@ -0,0 +1,4 @@
+<!-- Value - value of the search field (for search results page) -->
+<!-- Disabled (optional) - if search field has to be disabled -->
+<!-- Placeholder (optional) - placeholder text to be used -->
+<input type="search" spellcheck="false" name="q" maxlength="255" placeholder="{{if .Placeholder}}{{.Placeholder}}{{else}}{{ctx.Locale.Tr "search.search"}}…{{end}}"{{if .Value}} value="{{.Value}}"{{end}}{{if .Disabled}} disabled{{end}} autofocus>
```
if disabled 'disable' **else** 'autofocus'

silverwind commented:
I like unified style, but I'm not sure I like primary color on the search button. It calls too much attention imho for a secondary action.

silverwind commented on templates/shared/search/button.tmpl:
```diff
@@ -0,0 +1,3 @@
+<!-- Disable (optional) - if search button has to be disabled -->
+<!-- Tooltip (optional) - a tooltip to be displayed on hover -->
+<button class="ui small primary icon button" aria-label="{{ctx.Locale.Tr "search.search"}}" {{with .Tooltip}}data-tooltip-content="{{.}}"{{end}}{{if .Disabled}} disabled{{end}}>{{svg "octicon-search"}}</button>
```
What does `with` do and shouldn't that be an `if`?

denyskon replied:
If `.Tooltip` is set, the following code block will be executed and `.` becomes the value of `.Tooltip`. It's basically just a cleaner notation compared to the if-one and I use it pretty often. If you don't like it in Gitea I can also change it to if.

silverwind replied:
I guess `if` would be more common in gitea code. Changing the value of `.` is also often a problem as seen in loops, but I guess in this simple case it's ok.

denyskon replied:
I would like to keep `with` if it's not a blocker for you, as this one-line templates should stay as minimalistic as possible and this notation is simply shorter and easier to understand from my pov

silverwind replied:
I forbid `with` in my JS, but I guess for go template it might be borderline ok 😆

6543 replied:
I also had to look twice but i find it more elegant.

And if with is used more often people will notice imidiatly :)

denyskon commented:
I think our UI is getting more and more monochrome and it isn't a good thing. We already removed the color from the whole repo header, let's at least keep colored buttons..... I think the primary color is exactly right there

6543 commented:
As it is a refactor pull ... if this is the only issue to address i would argue to revert it as @silverwind suggested and after merge propose that change in a second pull

denyskon commented:
![grafik](https://github.com/go-gitea/gitea/assets/47871822/b18e53db-18a6-47c2-9cdc-313bb72fa35b)

Okay, I've reverted the primary color. But we really should do something against the greyscalification of our UI :laughing:

silverwind commented:
> But we really should do something against the greyscalification of our UI 😆

I see it as a evolution of the design. Take a look at other forges like GitLab or GitHub, there is almost no color left. I think it does give a bit more "professional" look.

6543 commented:
We should not discusst this out here ... but i personally woul find more collor beter to navigate etc ...

denyskon commented:
@silverwind Do you have any other concerns?

silverwind commented:
I have not reviewed in detail yet. But I recall @wxiaoguang had some reservations about this PR.

wxiaoguang commented:
> @wxiaoguang What exactly do you think is over-engineering here? :)

1. Personally I do not like the "{{if}}{{else}}" trick in templates.
2. This PR mixes too many different things together: "fuzzy" change, moved code, changed variable, changed logic
3. A lot of untested code, for example, at first glance, why `{{template "shared/search/input" dict "Value" .Query (ctx.Locale.Tr "search.package_kind")}}`? The "dict" accept key-value pairs, but I could only see 3 arguments there.

silverwind commented:
There is a height mismatch on /milestones page because of missing `small` class on `ui fluid action input`. `/issues` and `pulls` are okay but please check all occurences.

<img width="94" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/f235d3d2-62c7-430f-87ce-e11767ef425d">

silverwind commented:
Can something be done to make all search bars 35.4px? On `/explore` pages, it's 40px for example. Adding `small` there does not reduce the height.

wxiaoguang commented on templates/explore/user_list.tmpl:
```diff
@@ -26,6 +26,12 @@
 		</div>
 	{{else}}
-		<div class="flex-item">{{ctx.Locale.Tr "explore.user_no_results"}}</div>
+		<div class="flex-item">
+		{{if .PageIsExploreUsers}}
+			{{ctx.Locale.Tr "search.user_no_results"}}
+		{{else}}
+			{{ctx.Locale.Tr "search.org_no_results"}}
+		{{end}}
+		</div>
```
Could it be a general "search.no_results" ?

silverwind replied:
Or just `no_results` even. We need to cut down on unnecessary translation namespacing.

wxiaoguang commented on templates/repo/branch/list.tmpl:
```diff
@@ -67,140 +67,136 @@
 			</div>
 		{{end}}
 
-		{{if .Branches}}
```
The removed line seems to be a unrelated change. If I understand correctly, this PR's purpose is to "unify search boxes", but not "change the branch render logic" ....

denyskon replied:
The reason is that otherwise the search bar disappears when there are no results....

wxiaoguang commented on templates/repo/commits_table.tmpl:
```diff
@@ -8,27 +8,30 @@
+	{{end}}
 </h4>
 
+{{if .PageIsCommits}}
```
The old UI is done intentionally, to make it compact. The new UI makes the "input" a full line? Is it really good (and important) on the commit list page?

![image](https://github.com/go-gitea/gitea/assets/2114189/842e0129-94ab-4ed6-9b5c-eaba1171d4e2)

denyskon replied:
It's the way it is done on other tables, e. g. on all the admin pages. I think it is better so stay consistent, and I don't see a problem in one more line used.

silverwind replied:
Could move the search to the empty space outside the table in this case.

denyskon replied:
@silverwind I tried now, however I didn't manage to fix the styles involved :laughing:  I would leave it as done currently. This commits table should be reworked somewhen soon, and I can probably try to do it if you don't have time to continue your PR.

silverwind replied:
Feel free to, I don't think I will be motivated for that PR anytime soon 😆

wxiaoguang commented on templates/repo/home.tmpl:
```diff
@@ -15,14 +15,12 @@
-								<button class="ui small icon button"{{if .CodeIndexerUnavailable}} disabled{{end}} type="submit">
-									{{svg "octicon-search"}}
-								</button>
+								{{template "shared/search/button"}}
```
Missing `{{if .CodeIndexerUnavailable}}`?

denyskon replied:
```suggestion
								{{template "shared/search/button" dict "Disabled" .CodeIndexerUnavailable}}
```

wxiaoguang commented on templates/shared/repo_search.tmpl:
```diff
@@ -65,3 +65,4 @@
 		<span data-tooltip-content="{{ctx.Locale.Tr "explore.relevant_repositories_tooltip"}}">{{ctx.Locale.Tr "explore.relevant_repositories" (printf "?only_show_relevant=0&sort=%s&q=%s&language=%s" $.SortType (QueryEscape $.Keyword) (QueryEscape $.Language))}}</span>
 	</div>
 {{end}}
+<div class="divider"></div>
```
Why an extra "divider" here .... maybe it is done to make the UI look better, but maybe it could also not be considered as the purpose of "unifying search boxes"

denyskon replied:
It is present on the other pages like explore/users or explore/orgs. I could have moved it out to a separate PR, but I don't think it's necessary.

wxiaoguang commented on templates/shared/search/button.tmpl:
```diff
@@ -0,0 +1,3 @@
+<!-- Disable (optional) - if search button has to be disabled -->
+<!-- Tooltip (optional) - a tooltip to be displayed on hover -->
```
Could it use Golang template comment syntax? (here and others)

wxiaoguang commented on templates/shared/search/code/results.tmpl:
```diff
@@ -0,0 +1,44 @@
+<div class="flex-text-block gt-fw">
+	{{$pageURL := print $.SourcePath "/search"}}
+	{{if and $.PageIsExploreCode $.ContextUser}}
+		{{$pageURL = print AppSubUrl "/" $.ContextUser.Name "/-/code"}}
+	{{else if $.PageIsExploreCode}}
+		{{$pageURL = print AppSubUrl "/explore/code"}}
+	{{end}}
```
That's really tricky. These "if" blocks might be broken in the future and difficult to test.

Could it make the caller prepare the "PageURL" ahead?

denyskon replied:
I will check again what can be done with this template....

denyskon replied:
Turned out this template voodoo wasn't needed :laughing: Removed completely and replaced with a simple $.Link call

denyskon replied:
> The parent template, or the Go code?

 The parent template, or the Go code?

There is a `.Link` variable in every parent template that gives you the current relative URL as we always tass through the page object to that code search template, we can simply use it.

wxiaoguang commented on templates/shared/search/code/results.tmpl:
```diff
@@ -0,0 +1,44 @@
+</div>
+<div class="repository search">
+	{{range $result := .SearchResults}}
+		{{$repoLink := $.SourcePath}}
+		{{if $.PageIsExploreCode}}{{$repoLink = (index $.RepoMaps .RepoID).Link}}{{end}}
```
It also looks fragile here and below (by using `if`), that's what I meant "over-engineering" .....

wxiaoguang replied:
If there are so many different details, I think it's better to use 2 separate template files, instead of mixing the "if" logics into one template file.

denyskon replied:
SImplified just to decide between .Repo or $.RepoID in the beginning

wxiaoguang commented on templates/shared/search/input.tmpl:
```diff
@@ -0,0 +1,4 @@
+<!-- Value - value of the search field (for search results page) -->
+<!-- Disabled (optional) - if search field has to be disabled -->
+<!-- Placeholder (optional) - placeholder text to be used -->
+<input type="search" spellcheck="false" name="q" maxlength="255" placeholder="{{if .Placeholder}}{{.Placeholder}}{{else}}{{ctx.Locale.Tr "search.search"}}…{{end}}"{{if .Value}} value="{{.Value}}"{{end}}{{if .Disabled}} disabled{{else}} autofocus{{end}}>
```
Should "autofocus" always be applied? I think in some cases the search input shouldn't grab the focus ahead.

silverwind replied:
Also see [this](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autofocus#accessibility_concerns):

> Automatically focusing a form control can confuse visually-impaired people using screen-reading technology and people with cognitive impairments. When autofocus is assigned, screen-readers "teleport" their user to the form control without warning them beforehand.

denyskon replied:
Removed

wxiaoguang commented on templates/shared/search/input.tmpl:
```diff
@@ -0,0 +1,4 @@
+<!-- Value - value of the search field (for search results page) -->
+<!-- Disabled (optional) - if search field has to be disabled -->
+<!-- Placeholder (optional) - placeholder text to be used -->
+<input type="search" spellcheck="false" name="q" maxlength="255" placeholder="{{if .Placeholder}}{{.Placeholder}}{{else}}{{ctx.Locale.Tr "search.search"}}…{{end}}"{{if .Value}} value="{{.Value}}"{{end}}{{if .Disabled}} disabled{{else}} autofocus{{end}}>
```
Maybe the dots in `{{ctx.Locale.Tr "search.search"}}…` should be also put into the translation string, to make it translatable, just like "repo_kind", etc

wxiaoguang reviewed (commented):
A brief review without really testing. 

TBH, I would suggest one PR should only focus on major purpose (unless mixing something is not avoidable, or the changes are small enough and all in one module/package)

Mixing too many things together makes it difficult to review or test, and would cause conflicts with many PRs frequently.

https://github.com/go-gitea/gitea/blob/main/CONTRIBUTING.md#pull-request-format

![image](https://github.com/go-gitea/gitea/assets/2114189/6fa0f069-2b19-4b52-9ea1-56097cef862d)

wxiaoguang commented on options/locale/locale_en-US.ini:
```diff
@@ -155,6 +155,30 @@ filter.not_template = Not Template
+team_kind = Search teams...
+code_kind = Search code...
+code_no_results = No matching source code found.
+code_search_unavailable = Code search is currently not available. Please contact your site administrator.
```
For a large instance, "your" seems not a proper word.

For example, if GitHub crashes, GitHub users couldn't find "their" site administrator. 🤣 


So maybe it could be "Please contact the site administrator."

denyskon commented:
@silverwind The height should be okay now. It's a mix of fomantic paddings (where do these magic numbers come from :cry: ) and missing "small" class....

silverwind commented on web_src/css/form.css:
```diff
@@ -41,6 +41,11 @@ textarea,
 
+/* fix fomantic small dropdown having inconsistent padding with input */
+.ui.small.selection.dropdown {
+  padding: .67857143em 3.2em .67857143em 1em
```
```suggestion
  padding: .67857143em 3.2em .67857143em 1em;
```

silverwind commented:
Ok, I will test heights again. Yeah I know it's tricky. We should eventually just get rid of fomantic classes for search altogether, which will hopefully be easier now that so much is in shared templates.

silverwind commented:
Found tiny issue: Borders are not continous here. Checking for a fix...

<img width="451" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/58570e78-bcde-43f4-bcd3-0cf0abf393d1">

denyskon commented:
The problem isn't really search but dropdowns in all their variety. Ideally we would make an own dropdown module

silverwind commented:
Some CSS fixed done. This was a bit tricky.

<img width="1269" alt="Screenshot 2024-03-14 at 20 40 40" src="https://github.com/go-gitea/gitea/assets/115237/1a0f1738-6e9c-480f-83b0-9f9662413eec">

<img width="1262" alt="Screenshot 2024-03-14 at 20 40 28" src="https://github.com/go-gitea/gitea/assets/115237/8bbb9b7b-d352-40b0-bf5c-b4d6446ec364">
<img width="1252" alt="Screenshot 2024-03-14 at 20 40 22" src="https://github.com/go-gitea/gitea/assets/115237/2ff0b97b-55ae-4854-81f2-d82d909e589e">

silverwind commented:
Ideally we would find a suitable framework with headless components that we can style our own, like these, but they are React/Vue only unfortunately:

- https://www.radix-ui.com/
- https://headlessui.com/

I imagine there must be similar frameworks out there for Vanilla JS.

silverwind reviewed (approved):
https://github.com/go-gitea/gitea/pull/29530#discussion_r1525418103 would be nice, otherwise I'm happy.

denyskon commented on templates/repo/commits_table.tmpl:
```diff
@@ -8,27 +8,30 @@
+				<select class="ui small dropdown" name="all">
+					<option value="false">{{ctx.Locale.Tr "repo.commits.search_branch"}}</option>
+					<option {{if .All}} selected="selected"{{end}} value="true">{{ctx.Locale.Tr "repo.commits.search_all"}}</option>
+				</select>
```
Done

denyskon commented:
@silverwind @wxiaoguang I think I've addressed everything. Should be ready to merge.

silverwind commented:
Doing one final style fix :)

silverwind commented:
Fix done, it was the right border here not being blue:

<img width="431" alt="Screenshot 2024-03-14 at 22 39 22" src="https://github.com/go-gitea/gitea/assets/115237/2f152cbe-b0dd-4be7-92ed-5dd454e23a9e">

delvh reviewed (commented):
Apparently I was too slow

wxiaoguang commented on templates/admin/org/list.tmpl:
```diff
@@ -7,7 +7,26 @@
 			</div>
 		</h4>
 		<div class="ui attached segment">
-			{{template "admin/base/search" .}}
```
I can answer your question. That's caused by an over-abstraction in old days.

At that time, `admin/base/search` was designed to be a "shared template" and was expected to be used everywhere. But over time, that design didn't fit more and more cases, so the "shared template" usages were replaced by dedicated code block one by one, then, here is the only last one place which uses the "shared template" nowadays.
